app/src/main/cpp/skyline/gpu/memory_manager.h

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: MPL-2.0
+// Copyright © 2020 Skyline Team and Contributors (https://github.com/skyline-emu/)
+
+#pragma once
+
+#include <common.h>
+
+namespace skyline {
+    namespace constant {
+        constexpr u64 GpuPageSize = 1 << 16; //!< The page size of the GPU address space
+    }
+
+    namespace gpu::vmm {
+        /**
+        * @brief This enumerates the possible states of a memory chunk
+        */
+        enum ChunkState {
+            Unmapped, //!< The chunk is unmapped
+            Allocated, //!< The chunk is allocated but unmapped
+            Mapped //!< The chunk is mapped and a CPU side address is present
+        };
+
+        /**
+        * @brief This describes a chunk of memory and all of it's individual attributes
+        */
+        struct ChunkDescriptor {
+            u64 address; //!< The address of the chunk in the GPU address space
+            u64 size; //!< The size of the chunk in bytes
+            u64 cpuAddress; //!< The address of the chunk in the CPU address space (if mapped)
+            ChunkState state; //!< The state of the chunk
+
+            /**
+            * @brief Checks if the given chunk can be contained wholly within this chunk
+            * @param chunk The chunk to check
+            * @return If the given chunk can be contained wholly within this chunk
+            */
+            inline bool CanContain(const ChunkDescriptor& chunk) {
+                return (chunk.address >= this->address) && ((this->size + this->address) >= (chunk.size + chunk.address));
+            }
+
+            ChunkDescriptor(u64 address, u64 size, u64 cpuAddress, ChunkState state) : address(address), size(size), cpuAddress(cpuAddress), state(state) {}
+        };
+
+        /**
+        * @brief The MemoryManager class handles the mapping of the GPU address space
+        */
+        class MemoryManager {
+          private:
+            std::vector<ChunkDescriptor> chunkList; //!< This vector holds all the chunk descriptors
+
+            /**
+             * @brief This finds a chunk of the specified type in the GPU address space that is larger than the given size
+             * @param size The minimum size of the chunk to find
+             * @param state The state desired state of the chunk to find
+             * @return The first unmapped chunk in the GPU address space that fits the requested size
+             */
+            std::optional<ChunkDescriptor> FindChunk(u64 size, ChunkState state);
+
+            /**
+             * @brief This inserts a chunk into the chunk list, resizing and splitting as necessary
+             * @param newChunk The chunk to insert
+             * @return The base virtual GPU address of the inserted chunk
+             */
+            u64 InsertChunk(const ChunkDescriptor &newChunk);
```

Comment:
The argument could be named just `chunk`, I am aware this is the most pointless nitpick ever

Reply:
This breaks my iterator, which is also named chunk... I think I'll keep it as newChunk as it's a bit clearer in the function itself.

Reply:
You can name things differently in the definition and declaration but that's probably worse than just having it be `newChunk` universally.

Reply:
Yeah, I'll leave it as is